app/src/main/java/org/thoughtcrime/securesms/database/helpers/PreKeyMigrationHelper.java

```diff
@@ -0,0 +1,226 @@
+  private static final int    CURRENT_VERSION_MARKER = 2;
+
+  private static final String TAG = Log.tag(PreKeyMigrationHelper.class);
+
+  static boolean migratePreKeys(Context context, SQLiteDatabase database) {
+    File[]  preKeyFiles = getPreKeyDirectory(context).listFiles();
+    boolean clean       = true;
+
+    if (preKeyFiles != null) {
+      for (File preKeyFile : preKeyFiles) {
+        if (!"index.dat".equals(preKeyFile.getName())) {
+          try {
+            PreKeyRecord preKey = new PreKeyRecord(loadSerializedRecord(preKeyFile));
+
+            ContentValues contentValues = new ContentValues();
+            contentValues.put(OneTimePreKeyDatabase.KEY_ID, preKey.getId());
+            contentValues.put(OneTimePreKeyDatabase.PUBLIC_KEY, Base64.encodeBytes(preKey.getKeyPair().getPublicKey().serialize()));
+            contentValues.put(OneTimePreKeyDatabase.PRIVATE_KEY, Base64.encodeBytes(preKey.getKeyPair().getPrivateKey().serialize()));
+            database.insert(OneTimePreKeyDatabase.TABLE_NAME, null, contentValues);
+            Log.i(TAG, "Migrated one-time prekey: " + preKey.getId());
+          } catch (IOException | InvalidMessageException e) {
+            Log.w(TAG, e);
+            clean = false;
+          }
+        }
+      }
+    }
+
+    File[] signedPreKeyFiles = getSignedPreKeyDirectory(context).listFiles();
+
+    if (signedPreKeyFiles != null) {
+      for (File signedPreKeyFile : signedPreKeyFiles) {
+        if (!"index.dat".equals(signedPreKeyFile.getName())) {
+          try {
+            SignedPreKeyRecord signedPreKey = new SignedPreKeyRecord(loadSerializedRecord(signedPreKeyFile));
+
+            ContentValues contentValues = new ContentValues();
+            contentValues.put(SignedPreKeyDatabase.KEY_ID, signedPreKey.getId());
+            contentValues.put(SignedPreKeyDatabase.PUBLIC_KEY, Base64.encodeBytes(signedPreKey.getKeyPair().getPublicKey().serialize()));
+            contentValues.put(SignedPreKeyDatabase.PRIVATE_KEY, Base64.encodeBytes(signedPreKey.getKeyPair().getPrivateKey().serialize()));
+            contentValues.put(SignedPreKeyDatabase.SIGNATURE, Base64.encodeBytes(signedPreKey.getSignature()));
+            contentValues.put(SignedPreKeyDatabase.TIMESTAMP, signedPreKey.getTimestamp());
+            database.insert(SignedPreKeyDatabase.TABLE_NAME, null, contentValues);
+            Log.i(TAG, "Migrated signed prekey: " + signedPreKey.getId());
+          } catch (IOException | InvalidMessageException e) {
+            Log.w(TAG, e);
+            clean = false;
+          }
+        }
+      }
+    }
+
+    File oneTimePreKeyIndex = new File(getPreKeyDirectory(context), PreKeyIndex.FILE_NAME);
+    File signedPreKeyIndex  = new File(getSignedPreKeyDirectory(context), SignedPreKeyIndex.FILE_NAME);
+
+    if (oneTimePreKeyIndex.exists()) {
+      try {
+        InputStreamReader reader = new InputStreamReader(new FileInputStream(oneTimePreKeyIndex));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *reader*. The resource is referenced at lines 92, 93. The resource is closed at line 94. There are other execution paths that do not contain closure statements, for example, when JsonUtils.fromJson throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).